osquery/tables/system/linux/disk_encryption.cpp

```diff
@@ -79,31 +102,42 @@ void genFDEStatusForBlockDevice(const std::string& name,
     }
 
     r["type"] = osquery::join(items, "-");
-    encrypted_rows[name] = r;
     break;
   }
-
-    // If there's no good crypt status, check to see if we've already
-    // defined the parent_name. If so, inherit data from there. This
-    // works because the `SQL::selectAllFrom("block_devices")` is
-    // ordered enough. If that order proves inadequate, we may need
-    // to explicitly sort it.
+  // If there's no good crypt status, use the parent device's crypt status.
   default:
-    if (encrypted_rows.count(parent_name)) {
+    // Step through each parent once until we either reach
+    // the root device, or a device with valid encryption.
+    if (!parent_name.empty()) {
+      if (!encrypted_rows.count(parent_name)) {
+        if (!block_devices.count(parent_name)) {
+          queryBlockDevice(runSelectAll, parent_name, block_devices);
+        }
```

Comment:
Early return will help make this more clear

```suggestion
    if (parent_name.empty()) {
      r["encryption_status"] = kEncryptionStatusNotEncrypted;
      r["encrypted"] = "0";
      r["type"] = "";
      break;
    }

    // If we have a parent, let's use its status
    if (!encrypted_rows.count(parent_name)) {
      // If we don't have data yet, query the parent
      if (!block_devices.count(parent_name)) {
          queryBlockDevice(runSelectAll, parent_name, block_devices);
        }
```